jupyterhub_entrypoint/dbi/entrypoints.py

```diff
@@ -0,0 +1,364 @@
+
+import itertools
+
+from sqlalchemy.exc import IntegrityError
+from sqlalchemy.sql import insert, select, update, delete
+from sqlalchemy.sql.expression import label
+
+from jupyterhub_entrypoint.dbi.model import entrypoints, entrypoint_tags, tags
+
+async def create_entrypoint(
+    conn,
+    user,
+    entrypoint_name,
+    entrypoint_type,
+    entrypoint_data,
+    tag_names=[]
+):
+    """Create user entrypoint with optional tags.
+
+    Args:
+        conn            (AsyncConnection): SQLAlchemy asyncio connection proxy
+        user            (str): User name
+        entrypoint_name (str): User-assigned entrypoint name
+        entrypoint_type (str): Type of user entrypoint
+        entrypoint_data (dict): Contains type-specific metadata
+        tag_names       (list of str, optional): Names of tags to associate
+
+    Raises:
+        ValueError: If insertion of the entrypoint record fails.
+        ValueError: If one or more named tags do not exist.
+
+    """
+
+    # FIXME verify this rolls back if tagging below fails
+
+    statement = (
+        insert(entrypoints)
+        .values(
+            user=user,
+            entrypoint_name=entrypoint_name,
+            entrypoint_type=entrypoint_type,
+            entrypoint_data=entrypoint_data
+        )
+    )
+    try:
+        results = await conn.execute(statement)
+    except IntegrityError:
+        raise ValueError
+    entrypoint_id = results.inserted_primary_key.id
+
+    if not tag_names:
+        return
+
+    statement = select(tags).where(tags.c.tag_name.in_(tag_names))
+    results = await conn.execute(statement)
+    tag_ids = [r.id for r in results.fetchall()]
+    if len(tag_ids) != len(tag_names):
+        raise ValueError
+
+    statement = (
+        insert(entrypoint_tags)
+        .values(
+            [dict(entrypoint_id=entrypoint_id, tag_id=t) for t in tag_ids]
+        )
+    )
+    await conn.execute(statement)
+
+async def retrieve_one_entrypoint(conn, user, entrypoint_name):
+    """Retrieve data and tags for a user's entrypoint by name.
+
+    Args:
+        conn            (AsyncConnection): SQLAlchemy asyncio connection proxy
+        user            (str): User name
+        entrypoint_name (str): User-assigned entrypoint name
+
+    Returns:
+        dict: Contains entrypoint data and list of tag names
+
+    Raises:
+        ValueError: If no entrypoint named `entrypoint_name` is found.
+
+    """
+
+    statement = (
```

Comment:
The way I've typically used sqlalchemy is through the ORM where you don't craft a SQL-ish statement but use query() with filter() 
eg. 
```
session.query(Employee).filter_by(name='Alice')
                             .order_by(Employee.id)
                             .limit(5)
                             .all()
```

This may be more of a stylistic thing but it would be worth checking if this syntax works with a non-SQL backend like Mongo.

Reply:
Not leaving individual comments on each SQL invocation, so if you decide to refactor it should be across the board. I do get the sense that query/filter is more canonical.

Reply:
We are using the SQLAlchemy core API here, not the ORM one, so "select" is right.  Moreover the ORM API's query will be unified with the Core API's select syntax, with query becoming a "legacy" feature: https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-unified-with-core-select